Added pytest based tests for portworx framework

jitupawar22 opened the pull request:
For https://portworx.atlassian.net/browse/PTX-787
Following functionalities implemented-
    Install, Uninstall, upgrade, cleanup, check service health.
Note that the cleanup script is place holder in this commit.
We need to improve overall post uninstall cleanup for portworx.

Commit:
Added pytest based tests for portworx framework
Following functionalities implemented-
    Indtall, Uninstall, upgrade, cleanup, check service health.
Note that the cleanup script is place holder in this cimmit.
We need to improve overall post uninstall cleanup for portworx.

piyush-nimbalkar commented on frameworks/portworx/scripts/px_dcos_cleanup.sh:
```diff
@@ -0,0 +1,75 @@
+echo "Shutting down inactive services"
+echo "---------------------------------------------------------------------------";
+
+px_services=`dcos service --inactive | grep -v "ACTIVE  TASKS  CPU" | grep "portworx" | awk '{print $7}'`
```
Instead of this use jq to extract the ID
`dcos service --inactive --json | jq -r '.[] | select(.name=="portworx") | .id'`

piyush-nimbalkar commented on frameworks/portworx/tests/config.py:
```diff
@@ -1,2 +1,26 @@
+PX_CLEANUP_SCRIPT_PATH = 'frameworks/portworx/scripts/px_dcos_cleanup.sh'
+
+DEFAULT_TASK_COUNT = 3
+DEFAULT_CASSANDRA_TIMEOUT = 600
```
Remove this

piyush-nimbalkar commented on frameworks/portworx/tests/config.py:
```diff
@@ -1,2 +1,26 @@
+DEFAULT_TASK_COUNT = 3
+DEFAULT_CASSANDRA_TIMEOUT = 600
+
+DEFAULT_NODE_ADDRESS = os.getenv('PORTWORX_NODE_ADDRESS', sdk_hosts.autoip_host(SERVICE_NAME, 'node-0-server'))
```
Is this needed?
If yes let's use the vip host instead of pointing to zeroth node. Anyway `node-0-server` is incorrect , it should have been `portworx-0-install`

jitupawar22 replied:
Not now, but it will be needed in future.  Changing name to portworx-0-install

piyush-nimbalkar replied:
If you don't need to talk to a specific node, I would recommend using `vip_host(SERVICE_NAME, 'api')` which is like a load balancer for portworx nodes.

jitupawar22 replied:
Sure, Will consider using vip_host(SERVICE_NAME, 'api'). removing this now.

piyush-nimbalkar replied:
Remove it or change it to `vip_host`

piyush-nimbalkar commented on frameworks/portworx/tests/config.py:
```diff
@@ -1,2 +1,26 @@
+
+DEFAULT_NODE_ADDRESS = os.getenv('PORTWORX_NODE_ADDRESS', sdk_hosts.autoip_host(SERVICE_NAME, 'node-0-server'))
+
+PX_NODE_OPTIONS = { "node": { "portworx_options": "-a -x mesos -d enp0s8 -m enp0s8",
```
Can we take these individual options as env variables and combine them in the config?
At least kvdb servers should not be committed in the code.

piyush-nimbalkar commented on frameworks/portworx/tests/conftest.py:
```diff
@@ -0,0 +1,12 @@
+
+@pytest.fixture(scope='session')
+def configure_security(configure_universe):
+    yield from sdk_security.security_session('portworx')
```
Can we use the SERVICE_NAME instead of `portworx` here?

Commit:
Added tests for updating portworx installation
Tests updates px iinstallation with changing parameters from node, lighthouse, etcd etc. sections.

Commit:
Added post uninstall portworx cleanups related things
This is needed in order to support upgrade or reinstall portworx.

piyush-nimbalkar commented on testing/sdk_install.py:
```diff
@@ -173,6 +180,44 @@ def _portworx_cleanup():
+    exit_status = True
+    output_agent = ""
+
+    for i in range(len(agents)):
```
This is a vagrant specific cleanup routine and will not work for all deployments. For instance if someone wants to run a sanity test on AWS, this will not work.
May be take the username and key from env vars.

jitupawar22 replied:
Yeah good catch. We need to consider cleanups for all cluster types.

piyush-nimbalkar replied:
Create a ticket for this and put it in next PR.

jitupawar22 replied:
Created PXT-831

piyush-nimbalkar commented on testing/sdk_install.py:
```diff
@@ -173,6 +180,44 @@ def _portworx_cleanup():
+        log.info("PORTWORX: i==> {} Agent hostname: {}".format(i,  agents[i]['hostname']))
+        # Skipping exit status checks
+        cmd = 'sudo systemctl stop portworx'
+        exit_status, output_agent = shakedown.run_command_on_agent('192.168.65.131', cmd, 'vagrant','/ssh/key')
```
Use `agents[i]['hostname']`

piyush-nimbalkar commented on testing/sdk_install.py:
```diff
@@ -152,18 +152,25 @@ def uninstall(
-        'pxctl -j v l', 'vagrant','/ssh/key')
+    try:
+        exit_status, output_agent = shakedown.run_command_on_agent(agents[0]['hostname'],
+            'pxctl -j v l', 'vagrant','/ssh/key')
```
Same as below comment, don't hard code the username and ssh key.

piyush-nimbalkar commented on frameworks/portworx/tests/test_update_px_service.py:
```diff
@@ -0,0 +1,165 @@
+# Here are tests for variour update options available for portworx.
```
*various

piyush-nimbalkar commented on frameworks/portworx/tests/test_update_px_service.py:
```diff
@@ -0,0 +1,165 @@
+        sdk_install.uninstall(config.PACKAGE_NAME, config.SERVICE_NAME)
+
+@pytest.mark.sanity
+def test_update_service_node():
```
*test_update_node_count would be more precise

piyush-nimbalkar commented on frameworks/portworx/tests/test_update_px_service.py:
```diff
@@ -0,0 +1,165 @@
+    update_service_node_to_3(portworx_service)
+
+def update_service_node_to_3(portworx_service: dict):
+    update_options = {
```
nit:  It this method is just forming the options, you could combine it with the actual test method. Same for below tests.

piyush-nimbalkar commented on frameworks/portworx/tests/test_update_px_service.py:
```diff
@@ -0,0 +1,165 @@
+        options_path = f.name
+
+        log.info("Writing updated options to %s", options_path)
+        json.dump(options, f)
```
Here we are sending only the changed options. Does the update work without sending rest of the existing options? Does it auto merge?
Haven't tried it before myself.

jitupawar22 replied:
Yes, the update works without sending rest of the existing options, it does auto merge.

piyush-nimbalkar commented on frameworks/portworx/tests/test_sanity.py:
```diff
@@ -1,25 +1,48 @@
+# Verify portworx installation
+@pytest.mark.sanity
+def test_verify_install():
+    shakedown.service_healthy(config.SERVICE_NAME)
```
This seems to be checking that the tasks are healthy and in running state.
Can we also check status of portworx service and px status?

jitupawar22 replied:
That's right, we can additionally verify that as well.

piyush-nimbalkar replied:
We should verify that, otherwise it is not a complete test. We can't just rely on the framework install being successful, we have to check if Portworx cluster is up and running as well.

piyush-nimbalkar commented on testing/sdk_install.py:
```diff
@@ -260,7 +305,7 @@ def marathon_dropped_service():
             sdk_utils.list_reserved_resources()
     # Call portworx specific cleanup routine at the end.
-    _portworx_cleanup()
+    _portworx_volcleanup()
```
I feel we should call the portworx volume cleanup from the tests wherever we are uninstalling the framework, because it does not make much sense to do volume cleanup on portworx framework cleanup itself.

jitupawar22 replied:
To avoid calls from multiple places I considered calling it at the end of uninstall function itself. 
But I now I am also thinking that portworx specific cleanups should be outside from common framework cleanups.

piyush-nimbalkar commented on testing/sdk_install.py:
```diff
@@ -152,18 +152,25 @@ def uninstall(
-# Portworx specific cleanups are done here. 
-def _portworx_cleanup():
+# Portworx volume specific cleanups are done here. 
+def _portworx_volcleanup():
```
nit: `_portworx_vol_cleanup`?

piyush-nimbalkar commented on frameworks/portworx/tests/test_sanity.py:
```diff
@@ -1,25 +1,48 @@
+# Do post uninstall cleanup
+@pytest.mark.sanity
+def test_post_uninstall_cleanup():
+    assert sdk_install.portworx_cleanup() == 0
```
Instead of asserting on the return status, just raise an exception on failure just like other sdk functions (install, uninstall, etc)

piyush-nimbalkar replied:
Can you address this?

piyush-nimbalkar commented on frameworks/portworx/tests/test_sanity.py:
```diff
@@ -1,25 +1,48 @@
+def test_post_uninstall_cleanup():
+    assert sdk_install.portworx_cleanup() == 0
 
+# Upgrade portworx
```
*Upgrade portworx framework from released version

piyush-nimbalkar commented on frameworks/portworx/tests/test_update_px_service.py:
```diff
@@ -0,0 +1,165 @@
+    update_service(portworx_service, update_options)
+
+
+def update_service(service: dict, options: dict):
```
where do we use the service?

piyush-nimbalkar commented on frameworks/portworx/tests/test_update_px_service.py:
```diff
@@ -0,0 +1,165 @@
+        }
+    }
+
+    sdk_install.uninstall(config.PACKAGE_NAME, config.SERVICE_NAME)
```
Do we need to always reinstall the framework for changing options?
May be we can pass boolean which tells us to optionally uninstall? It will save us some time in test runs.

jitupawar22 replied:
No we don't need to uninstall every time. 
That uninstall don't get called every time. That gets call only first time before fresh installation.

piyush-nimbalkar replied:
The `portworx_service` method seems to be called from all the tests. How does this not get triggered every time?

jitupawar22 replied:
That is because we have marked portworx_service as  @pytest.fixture. It gets call only once for pytest session. And at teardown it invokes the cleanup code wrote in "finally:". 
Even if we don't call this function from any of the tests it will get called before executing any test from that file as the flag 'autouse=True' is used along with fixture.  I have made calls from every test such that user can have flexibility to run any individual test as well.

piyush-nimbalkar commented on frameworks/portworx/tests/test_sanity.py:
```diff
@@ -1,25 +1,48 @@
-def test_install():
-    pass # package installed and appeared healthy!
+def test_upgrade_framework():
+    sdk_install.uninstall(config.PACKAGE_NAME, config.get_foldered_service_name())
```
Looks like the `test_upgrade` method is already doing an uninstall of the service. We don't need to do it explicitly.

piyush-nimbalkar replied:
Do we need to need to the uninstall here? Isn't `test_upgrade` already doing it?

piyush-nimbalkar replied:
My bad, I see it is commented. Can you just remove the line?

piyush-nimbalkar commented on frameworks/portworx/tests/test_sanity.py:
```diff
@@ -1,25 +1,48 @@
-    pass # package installed and appeared healthy!
+def test_upgrade_framework():
+    sdk_install.uninstall(config.PACKAGE_NAME, config.get_foldered_service_name())
+    sdk_upgrade.test_upgrade(
```
Does this install the universe version and then upgrade to the `stub-universe` version?
When it does so, will the universe repo always remain on the top? Other tests that do an install, will they use the universe package or the stub-universe package?

jitupawar22 replied:
It use universe  package and then upgrade to stub-universe. 
Other tests from automation always use stub-universe.

Commit:
Made changes as per review comments.
Added support for passing px config parameters as environment variables.

Commit:
Added portworx image update test with provision to pass the image details to pytest framework
Also made changes as per review comments.
Purposefully not made changes related to ssh/key and volume_cleanup related stuffs.

Commit:
Missed in last checkin.
portworx_vol_function can be called from outside sdk_install.py on need basis.

piyush-nimbalkar commented on frameworks/portworx/tests/test_update_px_service.py:
```diff
@@ -0,0 +1,169 @@
+            }
+        }
+
+    update_service(update_options)
```
We should verify the specific changes that these options are going to bring in Portworx. For instance,
Verify the count of Portworx nodes has actually changed to 3.
Verify that portworx is actually upgraded to the given image version.
Verify lighthouse in enabled and reachable (do a curl and check)
Verify lighthouse is not running.
and so on

jitupawar22 replied:
The pytest sdk internally verify few things for example-
if we change node from 1 to 3 the test framework waits till 3 nodes comes up and healthy, if it fails to come up before timeout then it fails the test. 
same goes with lighthouse node. For upgrade image also it destroy all nodes and waits for installation with updated image on all nodes. If existing nodes does not destroy or nodes with new image does not come up then the framework fails the test. 
I agree that we should do portworx specific verification those not get covered from SDK.

jitupawar22 replied:
Updated PR with necessary verification those are not get addressed through pytest SDK.

piyush-nimbalkar commented on frameworks/portworx/tests/test_update_px_service.py:
```diff
@@ -0,0 +1,169 @@
+    update_service(update_options, False)
+
+@pytest.mark.sanity
+def test_update_enable_etcd():
```
Can you verify manually if changing the etcd is allowed once the framework is deployed? If yes what happens to Portworx in that case? Portworx should stop functioning because its kvdb itself changes.

jitupawar22 replied:
It is allowed to change etcd once the framework is deployed - Actually it should not allowed. 
Yes, manually I saw portworx stops functioning after changing etcd .

Commit:
Removed assert and used raise instead
some other cleanups as per review comments.
Will add portworx specific verifications in other PR.

Commit:
Implemented px_utils library for portworx specific functions
Added Portworx specific verifications
Added dcos portworx plan related testcases
Added dcos portworx pod related testcases

Commit:
Automated following tests-
    Suspend and resume portworx service
    Collect the configuration for the installed service and verify that it is as per provided configuration
    Replace portworx pod and move it to new agent
    List the pod instances and verify that those are as per deployment

piyush-nimbalkar commented on frameworks/portworx/tests/px_utils.py:
```diff
@@ -0,0 +1,85 @@
+def wait_for_plan_complete(plan_name):
+    count = 0
+    
+    while count < config.PX_TIMEOUT:
```
nit: Just do a time comparison here instead of keeping a count. Record a time before the while loop and check  
`start + PX_TIMEOUT < datetime.now()`

piyush-nimbalkar commented on frameworks/portworx/tests/px_utils.py:
```diff
@@ -0,0 +1,85 @@
+    cmd = "portworx status | jq '. | .Nodes | length'"
+    _, px_node_count, std_err = sdk_cmd.run_raw_cli(cmd)
+    
+    if True != is_int(px_node_count):
```
nit: simplify -
`if not is_int(px_node_count):`

piyush-nimbalkar commented on frameworks/portworx/tests/test_sanity.py:
```diff
@@ -27,6 +32,63 @@ def configure_package(configure_security):
+    px_status = px_utils.check_px_status() 
+   
+    log.info("Portworx service status is: {}".format(px_status))
+    if 2 != px_status:
```
Here you could use
`assert px_status == 2`

same for below tests

piyush-nimbalkar commented on frameworks/portworx/tests/test_sanity.py:
```diff
@@ -39,21 +39,76 @@ def test_verify_install():
+    if pod_agent_id_old == pod_agent_id_new:
+        log.info("PORTWORX: Failed to replace and move pod. Old pod agent id: {} , New pod agent id: {}".format(pod_agent_id_old, pod_agent_id_new))
+        raise
+# Verify suspend and resume portworx service
```
nit: new line here

piyush-nimbalkar commented on frameworks/portworx/tests/test_sanity.py:
```diff
@@ -45,6 +45,6 @@ def test_uninstall_package():
-    assert sdk_install.portworx_cleanup() == 0
-
-
+    if sdk_install.portworx_cleanup() != 0:
```
What I meant was that portworx_cleanup() method should raise an exception on failure instead of returning a code. Here the code will be just calling `sdk_install.portworx_cleanup()`

piyush-nimbalkar reviewed (approved):
This looks good to me. 
Address the minor comments and merge it in.

Commit:
Automated following tests-
    - Create px volume
    - update px volume
    - delete px volume
Made changes as per review comments.
Disabled etcd related tests from sanity.

Commit:
Missed in last commit, added assert

piyush-nimbalkar commented on frameworks/portworx/tests/px_utils.py:
```diff
@@ -127,14 +128,53 @@ def check_px_status():
+
+def px_get_vol_size_in_gb(vol_name):
+    gb_factor = 1024 * 1024 * 1024
+    cmd = "portworx volume list | jq '.[] | select(.locator.name == \"" + vol_name +"\") | .spec.size'" 
```
Could have done a volume inspect, right?

piyush-nimbalkar commented on frameworks/portworx/tests/px_utils.py:
```diff
@@ -127,14 +128,53 @@ def check_px_status():
     agent_id = px_get_agent_id(pod_name)
-    agent_id = str(agent_id)
-    cmd = "node ssh  \"sudo systemctl stop portworx\" --user=" + config.PX_AGENT_USER + " --mesos-id=" + agent_id
+    cmd = "node ssh  \"sudo systemctl stop portworx\" --user=" + config.PX_AGENT_USER + " --mesos-id=" + agent_id + " --option StrictHostKeyChecking=no"
```
could wrap this in a method which takes the agent user and agent id
eg. 
```
run(cmd, user=config.PX_AGENT_USER, agent_id=agent_id)
```
